File: today3/lru-cache.ts
/**
 * https://leetcode.com/problems/lru-cache/
 */
export { LRUCache, CacheItem };

class LRUCache {
  dummyHead: CacheItem;
  dummyTail: CacheItem;
  cache = new Map<number, CacheItem>();

  constructor(private capacity: number) {
    if (capacity <= 0) {
      throw Error(`LRUCache capacity must be a positive integer. Recieved: ${capacity}`);
    }

    this.dummyHead = {
      key: 'HEAD' as any,
      value: 'HEAD' as any,
      next: undefined as any,
      previous: undefined as any,
    };
    this.dummyTail = {
      key: 'TAIL' as any,
      value: 'TAIL' as any,
      next: undefined as any,
      previous: undefined as any,
    };
    this.dummyHead.next = this.dummyTail;
    this.dummyTail.previous = this.dummyHead;
  }

  get(key: number): number {
    const item = this.cache.get(key);
    if (item) {
      this.moveToFront(item);
      return item.value;
    }

    return -1;
  }

  put(key: number, value: number): void {
    let item = this.cache.get(key);
    if (item) {
      item.value = value;
      this.moveToFront(item);
    } else {
      if (this.cache.size === this.capacity) {
        // remove LRU
        this.cache.delete(this.dummyTail.previous.key);
        this.dummyTail.previous.previous.next = this.dummyTail;
        this.dummyTail.previous = this.dummyTail.previous.previous;
      }

      item = { key, value, next: this.dummyHead.next, previous: this.dummyHead };
      this.cache.set(key, item);
      this.dummyHead.next.previous = item;
      this.dummyHead.next = item;
    }
  }

  moveToFront(item: CacheItem) {
    if (item !== this.dummyHead.next) {
      item.previous.next = item.next;
      item.next.previous = item.previous;

      item.next = this.dummyHead.next;
      item.previous = this.dummyHead;

      this.dummyHead.next.previous = item;
      this.dummyHead.next = item;
    }
  }

  getValuesForward(): any[] {
    let pointer: CacheItem | undefined = this.dummyHead;
    const result: any[] = [];
    while (pointer) {
      result.push(pointer.value);
      pointer = pointer.next;
    }
    return result;
  }

  getValuesBackwards(): any[] {
    let pointer: CacheItem | undefined = this.dummyTail;
    const result: any[] = [];
    while (pointer) {
      result.push(pointer.value);
      pointer = pointer.previous;
    }
    return result;
  }
}

interface CacheItem {
  key: number;
  value: number;
  previous: CacheItem;
  next: CacheItem;
}

File: python/today12/lru_cache.py
# https://leetcode.com/problems/lru-cache/

from typing import Dict, List, Optional

class Node:
    def __init__(self, key = None, value = None, prev = None, next = None):
        self.key = key
        self.value = value
        self.prev = prev
        self.next = next
        
class LRUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.size = 0
        self.hash: Dict[int,Node] = {}

        self.head = Node('HEAD')
        self.tail = Node('TAIL')
        self.head.prev = self.tail
        self.tail.next = self.head

    def get(self, key: int) -> int:
        if key in self.hash:
            node = self.hash[key]
            self.move_to_tail(node)
            return node.value
        else:
            return -1

    def put(self, key: int, value: int) -> None:
        if key in self.hash:
            node = self.hash[key]
            node.value = value
            self.move_to_tail(node)
        else:
            if self.size < self.capacity:
                self.size += 1
            else:
                self.hash.pop(self.head.prev.key)
                new_head_prev = self.head.prev.prev
                self.head.prev = new_head_prev
                new_head_prev.next = self.head

            node = Node(key, value)
            self.hash[key] = node
            node.prev = self.tail
            node.next = self.tail.next
            self.tail.next.prev = node
            self.tail.next = node

    def move_to_tail(self, node: Node):
        if node.prev != self.tail:
            node.prev.next = node.next
            node.next.prev = node.prev
            node.prev = self.tail
            node.next = self.tail.next
            self.tail.next.prev = node
            self.tail.next = node

def to_array_forward(head: Optional[Node]) -> List[int]:
    if not head:
        return []
    result = []
    while head:
        result.append(head.key)
        head = head.next
    return result

def to_array_reverse(tail: Optional[Node]) -> List[int]:
    if not tail:
        return []
    result = []
    while tail:
        result.append(tail.key)
        tail = tail.prev
    return result